### PR TITLE
Refactor and clean up json avro schema converter

### PR DESCRIPTION
**Pull request by tuliren**
## What
- Merge record schemas within a type union.
- Rename methods and add Javadoc for clarity.
- Clean up test case.

### CODE REVIEW

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/util/AvroRecordHelper.java [hunk 1]
```diff
@@ -52,32 +47,4 @@ public static JsonNode pruneAirbyteJson(final JsonNode input) {
...
-      }
-    } else if (jsonNode.isArray()) {
-      ArrayNode arrayNode = (ArrayNode) jsonNode;
-
-      for (int i = 0; i < arrayNode.size(); i++) {
-        String arrayPath = currentPath + "/" + i;
-        obtainPaths(arrayPath, arrayNode.get(i), jsonNodePathMap);
-      }
-    }
-  }
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/util/AvroRecordHelper.java:81]: This method proactively calculate all paths to generate the namespaces. However, this is not necessary. The namespaces can be computed while travesing the Json schema.

File: airbyte-integrations/connectors/destination-s3/src/test/resources/parquet/json_schema_converter/json_conversion_test_cases.json [hunk 2]
```diff
@@ -1216,5 +1168,226 @@
       "identifier": ["151", "152", "true", "{\"id\":153}"],
       "_airbyte_additional_properties": null
     }
+  },
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-s3/src/test/resources/parquet/json_schema_converter/json_conversion_test_cases.json:1171]: The above case is trying to test objects with same names. However, it had lots of irrelevant fields. Those fields have been removed to make the test case easier to read.

File: docs/understanding-airbyte/json-avro-conversion.md [hunk 3]
```diff
@@ -161,6 +161,149 @@ This is not supported in Avro schema. As a compromise, the converter creates a u
 }
 ```
 
+If the Json array has multiple object items, these objects will be recursively merged into one Avro record. For example, the following Json array expects two different objects, each with a different `id` field.
```
Comment (sherifnada) [inline docs/understanding-airbyte/json-avro-conversion.md:164]: should this doc be in undertanding airbyte? or is it more appropriately scoped for the S3/blob storage connectors docs?

Comment (tuliren) [reply docs/understanding-airbyte/json-avro-conversion.md:164]: This avro doc was previously inline in the S3 and other blob storage connector docs. But as it gets longer and longer, I don't want to duplicate this doc in multiple places anymore. And In the future, more connectors using Avro / Parquet staging files will be related to this topic. So I think keeping everything here is the way to go. All relevant connector docs alreadt have links to this doc.

Comment (tuliren) [reply docs/understanding-airbyte/json-avro-conversion.md:164]: I realized that the wording is slightly off. I have updated the doc in a second Avro PR: #9367.

**Pull request merged**